zerver/lib/fetch.py

```diff
@@ -0,0 +1,474 @@
+        user = ctx.user
+        realm = user.realm
+        if user.is_guest:
+            state["realm_default_streams"] = []
+        else:
+            state["realm_default_streams"] = streams_to_dicts_sorted(
+                get_default_streams_for_realm(realm.id)
+            )
+
+    @staticmethod
+    def default_stream_groups(state: Dict[str, Any], ctx: FetchContext) -> None:
+        user = ctx.user
+        realm = user.realm
+        if user.is_guest:
+            state["realm_default_stream_groups"] = []
+        else:
+            state[
+                "realm_default_stream_groups"
+            ] = default_stream_groups_to_dicts_sorted(get_default_stream_groups(realm))
+
+    @staticmethod
+    def stop_words(state: Dict[str, Any], ctx: FetchContext) -> None:
+        state["stop_words"] = read_stop_words()
+
+    @staticmethod
+    def update_display_settings(state: Dict[str, Any], ctx: FetchContext) -> None:
+        user = ctx.user
+        for prop in UserProfile.property_types:
+            state[prop] = getattr(user, prop)
+        state["emojiset_choices"] = user.emojiset_choices()
+
+    @staticmethod
+    def update_global_notifications(state: Dict[str, Any], ctx: FetchContext) -> None:
+        user = ctx.user
+        for notification in UserProfile.notification_setting_types:
+            state[notification] = getattr(user, notification)
+        state["available_notification_sounds"] = get_available_notification_sounds()
+
+    @staticmethod
+    def user_status(state: Dict[str, Any], ctx: FetchContext) -> None:
+        user = ctx.user
+        realm = user.realm
+        state["user_status"] = get_user_info_dict(realm_id=realm.id)
+
+    @staticmethod
+    def video_calls(state: Dict[str, Any], ctx: FetchContext) -> None:
+        user = ctx.user
+        state["has_zoom_token"] = user.zoom_token is not None
+
+
+KNOWN_EVENT_TYPES = [
+    # The order here is historical, and it should not actually
+    # matter, since none of our fetcher have side effects.
+    # The names here should exactly match the set of static
+    # methods on Fetcher, and we possible want to do some fancy
+    # introspection in the future.
+    "version_info",
+    "alert_words",
+    "custom_profile_fields",
+    "hotspots",
+    "message",
+    "muted_topics",
+    "presence",
+    "realm",
+    "realm_domains",
+    "realm_emoji",
+    "realm_filters",
+    "realm_user_groups",
+    "realm_user",
+    "realm_bot",
+    "realm_embedded_bots",
+    "realm_incoming_webhook_bots",
+    "recent_private_conversations",
+    "subscription",
+    "update_message_flags",
+    "starred_messages",
+    "stream",
+    "default_streams",
+    "default_stream_groups",
+    "stop_words",
+    "update_display_settings",
+    "update_global_notifications",
+    "user_status",
+    "video_calls",
+]
+
+
+def fetch_initial_state_data(
+    user_profile: UserProfile,
+    event_types: Optional[Iterable[str]],
```

Comment:
Would only `Iterable[str]` be enough or do we have to have this `Optional`?

Reply:
Ah! `event_type` can be None when the user wants all events, so it's needed indeed.

Reply:
Good question!  This particular piece of code was just copy/paste on my part, so I didn't think about it, but trying to eliminate `Optional` where we can is a good goal.

Reply:
In this case `event_types = None` has a different meaning (include all fields) than the natural behavior of `event_types = []` (include no fields). So allowing a separate `None` is appropriate.